mktrayicon.c

```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
 #endif
 			break;
 		case 'm': /* menu */
-			fprintf(stderr, "Menu command not yet implemented\n");
-			if (param != NULL) {
-				free(param);
+			if (onmenu != NULL){
+				//destroy the previous menu
+				for (int i = 0; i < menusize; i++){
+					free(onmenu[i].name);
+					free(onmenu[i].action);
+					onmenu[i].name = NULL;
+					onmenu[i].action = NULL;
+				}
+				free(onmenu);
+				onmenu = NULL;
+				gtk_widget_destroy(menu);
+				menu = NULL;
 			}
+			
+			menusize = 0;
+
+			if (param != NULL && *param == '\0') {
+#ifdef DEBUG
+				printf("Removing onmenu handler\n");
+#endif
+				break;
+			}
+
+			// This block makes sure that the parameter after 'm' is well defined
+			int commas = 0;
+			int bars = 0;
+			if (len < 3){
+				break;
+			}
+			if (param[0] == ',' || param[0] == '|'){
+				printf("Action string cannot begin with ',' or '|'. Use '\\' to escape\n");
+				break;
+			}
+			if (param[len-2] == '|' && param[len-3] != '\\'){
+				printf("'|' can't be second of last. Use '\\' to escape\n");
+				break;
+			}
+			if (param[len-1] == ',' && param[len-2] != '\\'){
+				printf("',' can't be last. Use '\\' to escape\n");
+				break;
+			}
+			int i = 1; 
+			while (i < len - 1){
+				if (commas > bars + 1 || bars > commas)
+					break;
+				if (param[i] == ',' && param[i-1] != '\\')
+					commas++;
+				else if (param[i] == '|' && param[i-1] != '\\')
+					bars++;
+				i++;
+			}
+			if (commas > bars + 1 || bars > commas){
+				printf("Unbalanced ',' or '|'. Use '\\' to escape\n");
```

Comment:
Isn't it easier to just detect this when parsing below instead?

Reply:
At first glance I thought your point made sense. But now that I checked it again, if I make the checks simultaneously while parsing, I may have to free already allocated memory for the `onmenu` array

Reply:
It probably makes sense to pre-count the number of non-escaped `,`, that is true :)